server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/step/LoadReportAnalysisMetadataHolderStep.java

```diff
@@ -88,9 +88,10 @@ public void execute() {
 
     loadMetadata(reportMetadata);
     Organization organization = loadOrganization(reportMetadata);
-    loadProject(reportMetadata, organization);
+    Runnable projectValidation = loadProject(reportMetadata, organization);
     loadQualityProfiles(reportMetadata, organization);
     branchLoader.load(reportMetadata);
+    projectValidation.run();
```

Comment:
Why not calling `loadQualityProfiles(reportMetadata, organization);branchLoader.load(reportMetadata);` inside `loadProject()` ? That would avoid this non-conventional Runnable to reuse local variables.

Reply:
I agree with Simon. The `loadProject` method does validation in two stages. First when called, and second with the returned `Runnable`. This is really hacky, there's gotta be a better way to write it.